test/server/server_stats_flush_benchmark_test.cc

```diff
@@ -0,0 +1,73 @@
+#include <cstdint>
+#include <memory>
+
+#include "envoy/stats/sink.h"
+#include "envoy/stats/stats.h"
+
+#include "source/common/stats/thread_local_store.h"
+#include "source/server/server.h"
+
+#include "test/mocks/stats/mocks.h"
+#include "test/test_common/simulated_time_system.h"
+
+#include "absl/strings/str_cat.h"
+#include "benchmark/benchmark.h"
+#include "gmock/gmock.h"
+
+namespace Envoy {
+class StatsSinkFlushSpeedTest {
+
+public:
+  StatsSinkFlushSpeedTest(uint64_t n_counters, uint64_t n_gauges, uint64_t n_text_readouts)
+      : pool_(symbol_table_), stats_allocator_(symbol_table_), stats_store_(stats_allocator_) {
+
+    sinks_.emplace_back(new testing::NiceMock<Stats::MockSink>());
+    // Create counters
+    for (uint64_t idx = 0; idx < n_counters; ++idx) {
+      auto stat_name = pool_.add(absl::StrCat("counter.", idx));
+      stats_store_.counterFromStatName(stat_name).inc();
+    }
+    // Create gauges
+    for (uint64_t idx = 0; idx < n_gauges; ++idx) {
+      auto stat_name = pool_.add(absl::StrCat("gauge.", idx));
+      stats_store_.gaugeFromStatName(stat_name, Stats::Gauge::ImportMode::NeverImport).set(idx);
+    }
+
+    // Create text readouts
+    for (uint64_t idx = 0; idx < n_text_readouts; ++idx) {
+      auto stat_name = pool_.add(absl::StrCat("text_readout.", idx));
+      stats_store_.textReadoutFromStatName(stat_name).set(absl::StrCat("text_readout.", idx));
+    }
+  }
+
+  void test(benchmark::State& state) {
+    for (auto _ : state) {
+      UNREFERENCED_PARAMETER(state);
+      Server::InstanceUtil::flushMetricsToSinks(sinks_, stats_store_, time_system_);
+    }
+  }
+
+private:
+  Stats::SymbolTableImpl symbol_table_;
+  Stats::StatNamePool pool_;
+  Stats::AllocatorImpl stats_allocator_;
+  Stats::ThreadLocalStoreImpl stats_store_;
+  std::list<Stats::SinkPtr> sinks_;
+  Event::SimulatedTimeSystem time_system_;
+};
+
+static void bmLarge(benchmark::State& state) {
+  uint64_t n_counters = 1000000, n_gauges = 1000000, n_text_readouts = 1000000;
```

Comment:
s/n_counters/num_counters/ here and elsewhere in the PR

Actually it'd be better to parameterize this test on this number. Look at other "*_speed_test.cc" files for examples. Then you'd need only one test method, and you could parameterize a wider spectrum of values, so you could see where the knee of the curve was more easily than having just 2 data points.